lib/charms/observability_libs/v0/kubernetes_service_patch.py

```diff
@@ -67,6 +70,25 @@ def __init__(self, *args):
     # ...
 ```
 
+If you wish to specify the protocols for the corresponding service ports, use the protocols keyword
+argument. The protocols keyword argument is a list of strings. Valid protocols are "TCP", "UDP",
+and "SCTP". The length of the protocols list must be equal to the length of the ports list. If the
+protocols keyword argument is not used, the default protocol is TCP.
+```python
+# ...
+from charms.observability_libs.v0.kubernetes_service_patch import KubernetesServicePatch
+
+class SomeCharm(CharmBase):
+  def __init__(self, *args):
+    # ...
+    self.service_patcher = KubernetesServicePatch(
+        self,
+        [("svc1", 123, 123), ("svc2", 124, 124), ("svc3", 125, 125), ("svc4", 126, 126)],
+        protocols=["TCP", "UDP", "SCTP", "TCP"],
```

Comment:
Hmm, I'm a little uncomfortable with this - it feels a little gross to have to match the placement of the defined ports with this array.

I think this lib has grown specifically that we *could* consider moving to a new `v1` API version and break the old? @gruyaume any thoughts? I'd be interested to see what that might look like. Perhaps defining the services as a dict rather than a tuple?

@simskij this would require a change to bunch of your charms, although not *much* of a change, and only when the developers choose to bring the new library version in.

Reply:
We could NOT have a separate array for the ports, but accept ports in an extended format as `<protocol>/<port>`. Ports specified only as an int (or string that converts to an int) are assumed to be http for backwards compat.

Reply:
Fundamentally we have PortDefinition already specified for us to extend as we like. Adding the protocol there, either as part of the port value of (maybe better) as another element of the tuple, sounds the best option to me.

About PortDefinition: high time it becomes a DataClass or similar. Having it defined as a Union of tuples is pretty gross :D

Reply:
Okay so supporting `"80/TCP"` and `80` seems fine to me - we can just assume TCP as the default

Reply:
And agree on PortDefinition - when I originally implemented this it was much simpler!

Reply:
But then only in the source port? What would happen is source and target port have to different protocols? I gave it some more thought and I am thinking that having a separate protocol field in PortDefinition might be cleaner and less surprising (that’s how K8s does it after all).

Reply:
I'd agree with @mmanciop to create a dataclass for ports. Something like this would be pretty clean.


```
@dataclass
class Port:
    name: str
    port: int
    target_port: int = None
    node_port: int = None
    protocol: str = "TCP"
```

Reply:
Or perhaps at this point we just use the ServicePort from lightkube itself?! https://gtsystem.github.io/lightkube-models/1.17/models/core_v1/#serviceport

Reply:
@stonepreston up for giving it a shot?

Reply:
sure thing. Should be able to get the PR updated in the next few hours. I like the idea of just using the service port from lightkube, but if you can think of a reason to create a custom class I can do that as well.

Reply:
Seems like it would work pretty nicely to me -- @gruyaume/@mmanciop?

Reply:
Sounds good to me!

Reply:
Newest commit has the ServicePort changes